### DISCUSSION TITLE
Tips on speeding up in linux.git?

### DISCUSSION THREAD

**Discussion by troglobit** [gitext-rs/git-stack#216] [Q&A]
Hi, awseome little extension this! :sunglasses::+1:

We (me and @wkz) are playing around with it for stacking patch-sets (in the form of stacked git branches) in the Linux kernel tree.  Works great ... except for the time to run `git stack` ... in particular `git stack --repair` after cherry-picking a few commits to an intermediate branch.  Commands take up to a minute to complete.

Any tips on how to speed this up?

**Comment by epage** [comment]
What version are you running?  On the 16th, I released 0.7 which has massive speed improvements.  `git stack` on `gecko-dev` went from ~25s to less than 1s.

If you are on on 0.7 or newer, could you provide a repo I could clone with reproduction steps so I can profile it and see how we can improve it?

**Reply by wkz** [reply]
Wow, that's an impressive response time :smile: 

    $ git stack --version
    git-stack 0.8.0

I've pushed the relevant branches to https://github.com/westermo/linux/. I observe the following on my system:

    (wmo/linux-5.17.y-backports) $ time git stack
    ...lots of commit output...
    real    0m24.307s
    user    0m24.116s
    sys     0m0.188s

This on a Xeon E3-1230 with 32GB of RAM.

**Reply by epage** [reply]
I couldn't reproduce the slow down with just `master`, so I checked out each of the branches in that repo and am still not getting any slow down
```console
$ time git stack
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `/home/epage/src/personal/git-stack/target/debug/git-stack`
WARN: Branches contain more than 50 commits (should these be protected?): linux-5.17.ym net-bridge-host-l2-multicastm wmo/linux-5.17.y-backports
master (no remote) Revert "gpio: Revert regression in sysfs-gpio (gpiolib.c)"
⌽ 551acdc Merge tag 'net-5.17-final' of git://git.kernel.org/pub/scm/linux/kernel/git/netdev/net
├─┐
│ ⌽ net-bridge-host-l2-multicast (no remote) selftests: forwarding: verify flood of known mc on mcast_router port
⌽ linux-5.17.y (no remote) Linux 5.17
⌽ wmo/linux-5.17.y-backports (no remote) net: dsa: mv88e6xxx: Fill in STU support for all supported chips
Stack from other users: wmo/linux-5.17.y

real    0m0.658s
user    0m0.625s
sys     0m0.032s
```
Note that I'm getting extra output / slowdown because I have `git stack` aliased to `cargo run` the code straight from `main`.

**Reply by troglobit** [reply]
Admittedly, this is my first foray into anything written in Rust, so forgive my ignorance.  I managed to build and run the latest `main`, then ran it on my local branch (not pushed).  Tobias' `wmo/linux-5.17.y-backports` and `wmo/linux-5.17.y` are worse. 

```
joachim@wbg:~/src/linux.net-next [bridge-bum-flooding]$ time cargo run --manifest-path ~/src/git-stack/Cargo.toml 
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `/home/joachim/src/git-stack/target/debug/git-stack`
8dc84dc net: phy: broadcom: Enable 10BaseT DAC early wake
⌽ master Merge branch 'too-short'
├─┐
│ ⌽ b3b035b net: bridge: add control of bum flooding to bridge itself
│ ⌽ 74f1d8d net: bridge: rename br_switchdev_set_port_flag() to .._dev_flag()
│ ⌽ 85bd0e6 net: bridge: minor refactor of br_setlink() for readability
│ ⌽ f4e0419 net: bridge: netlink support for controlling BUM flooding to bridge
│ ⌽ bb52509 net: bridge: avoid classifying unknown multicast as mrouters_only
│ ⌽ ba59ec8 net: bridge: add support for host l2 mdb entries
│ ⌽ c69df56 selftests: forwarding: add TCPDUMP_EXTRA_FLAGS to lib.sh
│ ⌽ 5226ad2 selftests: extend tcpdump helper to handle multiple instances
│ ⌽ 32b2cb5 selftests: forwarding: new test, verify host mdb entries
│ ⌽ e1d6e03 selftests: forwarding: rename test groups for next bridge mdb tests
│ ⌽ f9a1a50 selftests: forwarding: verify flooding of unknown multicast
│ ⌽ 4c0c5ec selftests: forwarding: verify strict mdb fwd of known multicast
│ ⌽ f4257c4 selftests: forwarding: verify strict filtering doesn't leak
│ ⌽ net-bridge-host-l2-multicast (ready) selftests: forwarding: verify flood of known mc on mcast_router port
⌽ 87d7a42 net: bridge: add support for host l2 mdb entries
⌽ bridge-host-l2-mdb-entries (ready) selftests: forwarding: new test, verify host mdb entries
⌽ 81e23dc net: bridge: add control of bum flooding to bridge itself
⌽ 94d490f net: bridge: rename br_switchdev_set_port_flag() to .._dev_flag()
⌽ 26c6cbe net: bridge: minor refactor of br_setlink() for readability
⌽ b484971 net: bridge: netlink support for controlling BUM flooding to bridge
⌽ 440df2a selftests: forwarding: add TCPDUMP_EXTRA_FLAGS to lib.sh
⌽ 56827dbe selftests: extend tcpdump helper to handle multiple instances
⌽ bridge-bum-flooding selftests: forwarding: rename test groups for next bridge mdb tests
Stacks older than 14days: bridge-drop-initial-multicast-join-delay

real	0m14.813s
user	0m14.660s
sys	0m0.152s
```

I'm on Linux Mint 20.3 with `git --version` = 2.25.1 ... Tobias is on Ubuntu 20.04, which is the base for my Mint installation, so shared components so far.   After cargo update I got `libgit2-sys v0.13.1+1.4.2`, if that means anything.

**Reply by epage** [reply]
Oh, sorry, I wasn't expecting you to have to run off of `main` (there are rarely unreleased changes).

**EDIT:** Jump to the later edit if you want to find out root cause and paths forward (I just feel bad deleting all of this stuff I wrote up)

The challenge is the performance is dependent on the branches you have on your system.  For the branches I have access to, everything is fast.  If you don't mind making your local branches public, I can pull them and see whats going on.

As an alternative (since you have the code locally), you could [profile it](https://nnethercote.github.io/perf-book/profiling.html) and report back what you found.

That said, I do have one guess as to a slow down
```
8dc84dc net: phy: broadcom: Enable 10BaseT DAC early wake
⌽ master Merge branch 'too-short'
...
Stacks older than 14days: bridge-drop-initial-multicast-join-delay
```

We collapse old branches by default but this is purely cosmetic.  We'll first load the entire  commit graph for all of the branches and then we trim off the old branch.  You can see a vestige of this because there is a commit before `master`.  That is most likely the base commit between `master` and `bridge-drop-initial-multicast-join-delay`.  If its too old, I could see that slowing things down.

A quick way of verifying this, with the help if [git-branch-stash](https://github.com/gitext-rs/git-branch-stash)
```console
$ time git stack
$ git branch-stash
$ git branch -D bridge-drop-initial-multicast-join-delay
$ time git stack
$ git branch-stash pop
$ time git stack
```

**EDIT:** I realized I could verify this locally via
```console
$ time git stack
$ git checkout 8dc84dc
$ git checkout -b foo
$ gvim COPYING
$ git commit -a -m "foo"
$ time git stack
$ git co master
$ git branch -D foo
$ time git stack
```
Yes, the problem is how much of the commit graph we load because of `bridge-drop-initial-multicast-join-delay`

Some options (on both ends):
- You could delete the branch
- If there is a release branch (unsure if git uses those) prior to `bridge-drop-initial-multicast-join-delay`, you can check it out locally (make sure it is marked as "protected") and then `git-stack` will calculate `bridge-drop-initial-multicast-join-delay` relative to that release branch rather than `master`
  - Alternatively, you could fake this by creating your own branch at the merge base and mark it as protected
- `git stack` could switch from checking every unprotected commit in the branch to see if its too old to checking the tip of each branch and cutting them out earlier in the process (ie hide it earlier by skipping all of the steps that need the full commit graph loaded)

Unless someone has a better idea, I feel like cutting out old branches earlier in the processing is the best route forward as it makes things fast "out of the box".

**I'm going to give this some breathing room before moving forward with the above optimization to give everyone involved time to think about this and post any concerns or alternatives**

**Reply by arxanas** [reply]
What exactly is loaded as part of the commit graph, and what is it used for? git-branchless used to cache merge-base queries on disk; would that help? I suppose the slow part is for the ahead/behind calculation?

**Reply by epage** [reply]
I'm not double checked but my guess is its still `revwalk`.  I cheated in my last changes and went ahead and left in the topological sorts.  I just switched to l.ibgit2 deciding the terminating case, rather than git-stack.  This means it needs to do the topological sort on fewer commits, making it faster.  If enough history gets loaded, then thats still a problem.

The reason we have the topological sort is  we currently cheat and don't only track one parent for each commit (git stack generally discourages merge commits within topic branches and hides protected branch commits) so we need the commits linearized in a reasonable way.  One day, I'll have to actually go in and implement support for multiple parents.

**Reply by troglobit** [reply]
> Oh, sorry, I wasn't expecting you to have to run off of `main` (there are rarely unreleased changes).

No problem. I got to learn a few things, on a Monday. So all good :)
 
> **EDIT:** Jump to the later edit if you want to find out root cause and paths forward (I just feel bad deleting all of this stuff I wrote up)

OK

> [snip]
> **EDIT:** I realized I could verify this locally via
> ```
> $ time git stack
> $ git checkout 8dc84dc
> $ git checkout -b foo
> $ gvim COPYING
> $ git commit -a -m "foo"
> $ time git stack
> $ git co master
> $ git branch -D foo
> $ time git stack
> ```
> Yes, the problem is how much of the commit graph we load because of `bridge-drop-initial-multicast-join-delay`

Aha, I didn't realize that old "wip" branch, based off of Linux net-next master from a few months ago, could slow down the branch I was on. That's probably what hit @wkz as well, I'm sure he even has more such stale branches than I do.

> Some options (on both ends):
>     * You could delete the branch

Well, yeah but that's a bit harsh ... the commit freq on net-next is quick and I'm sure this is not unique to mine or @wkz workflows.  

>     * If there is a release branch (unsure if git uses those) prior to `bridge-drop-initial-multicast-join-delay`, you can check it out locally (make sure it is marked as "protected") and then `git-stack` will calculate `bridge-drop-initial-multicast-join-delay` relative to that release branch rather than `master`
>       * Alternatively, you could fake this by creating your own branch at the merge base and mark it as protected

Admittedly, I did not read the full documentation, so "protected" I never got to, sorry :-/

>     * `git stack` could switch from checking every unprotected commit in the branch to see if its too old to checking the tip of each branch and cutting them out earlier in the process (ie hide it earlier by skipping all of the steps that need the full commit graph loaded)

Sounds good to me.  I'm living in fantasy land wrt to git mostly, relying on pros like you to tell me what to do :-)
 
> Unless someone has a better idea, I feel like cutting out old branches earlier in the processing is the best route forward as it makes things fast "out of the box".

:+1:

Thank you so much for taking the time to respond and explain! <3

**Reply by epage** [reply]
> Aha, I didn't realize that old "wip" branch, based off of Linux net-next master from a few months ago, could slow down the branch I was on. 

atm the default is for `git stack` to operate on all branches.  [`stack.stack`](https://github.com/epage/git-stack/blob/main/docs/reference.md#config-fields) allows you to control this.  Try `git stack --stack dependents` to select your branch and all branches off of it.   You can then set it inside your `.gitconfig` if you want.  This sped up `git stack` from 18s to 5s for me.

`descendants` would show every branch off of `master` but all of your branches are off of `master` since there aren't other protected branches.

> Admittedly, I did not read the full documentation, so "protected" I never got to, sorry :-/

Its what `git stack` calls any branch that you do not want it to edit, whether because its a release branch or just another user's branch you were looking over.  It has a built-in list of common protected branches.  You can see this with `git stack --protected` and `git stack --dump-config -`.

**Reply by epage** [reply]
Just a note from my own investigation.  `git stack --stack dependents` is mostly spending its time on finding the merge-base between the ancient branch and everything else

Specifically, its these two.  The first is significantly slower than the second (caching?)
```
[TRACE git_stack::git::repo] merge 8335433a81cf6a62c7f9375886ae0a5d0da614f2 f443e374ae131c168a065ea1748feac6b2e76613
[TRACE git_stack::git::repo] merge -> Some(8dc84dcd7f74b50f81de3dbf6f6b5b146e3a8eea)
[TRACE git_stack::git::repo] merge 8335433a81cf6a62c7f9375886ae0a5d0da614f2 56e337f2cf1326323844927a04e9dbce9a244835
[TRACE git_stack::git::repo] merge -> Some(8dc84dcd7f74b50f81de3dbf6f6b5b146e3a8eea)
```

>  git-branchless used to cache merge-base queries on disk; would that help?

@arxanas I'm curious, how does `git-branchless` handle situations like this if it doesn't do on-disk caching?

**Reply by wkz** [reply]
Thanks for the great support!

I've done some more testing. It seems like performance degrades a lot when you have multiple protected branches that are "far apart". I.e. if I only have recent branches in my tree, `git stack` is _really_ fast:

```
(master) $ git br
  linux-5.17.y
* master
  wmo/linux-5.17.y
  wmo/linux-5.17.y-backports
(master) $ time git stack >/dev/null

real    0m0.023s
user    0m0.006s
sys     0m0.018s
```

Now if I add the stable branch of the previous release into the mix:

```
(master) $ git co linux-5.16.y 
Updating files: 100% (10249/10249), done.
Branch 'linux-5.16.y' set up to track remote branch 'linux-5.16.y' from 'stable'.
Switched to a new branch 'linux-5.16.y'
(linux-5.16.y) $ time git stack >/dev/null

real    0m2.104s
user    0m1.972s
sys     0m0.132s

```

The time increases almost 100x. Note that the newly added branch is protected:

```
(linux-5.16.y) $ git stack --protected
linux-5.16.y
linux-5.17.y
master
```

Adding an even older branch exacerbates the issue:

```
 (linux-5.16.y) $ git co linux-5.10.y 
Updating files: 100% (36486/36486), done.
Branch 'linux-5.10.y' set up to track remote branch 'linux-5.10.y' from 'stable'.
Switched to a new branch 'linux-5.10.y'
(linux-5.10.y) $ time git stack >/dev/null
Empty stacks: linux-5.16.y

real    0m4.900s
user    0m4.708s
sys     0m0.192s
```

**Reply by epage** [reply]
Yes, that looks to be about the slow down I would expect for adding protected branches.

There are a couple of ways of git-stack slowing down with different magnitudes. 
- Distant protected branches: this has the performance hit of calculating the merge-base between each branch and the protected branches
  - 0.8.1 does some caching but it won't always help in these cases.
  - Probably on-disk caching is the only way we can speed this up unless there is something special git-branchless is doing that we can do
  - We might also be able to do some inferring (If X is the common commit between A and B and B is the common commit between B and C, then X is the common commit of between A and C)
- Enumerating commits between a merge base and both the associated protected branch and the topic branches.
  - **This is much more expensive than having a protected branch** because we are doing a topological sort of all of the commits
  - We at least reuse calculations between a base and each of the branches
  - Options
    - Ignoring stale branches earlier in the process so the base commit is much closer
    - Do a major re-work of the internals to remove the topological sort

And thank you for being willing to push git-stack on performance and working with us to get it faster!  Its hard to scale to up in repo size when I'm not working on those sizes of repos of late.

**Reply by arxanas** [reply]
> I'm curious, how does git-branchless handle situations like this if it doesn't do on-disk caching?

@epage it now uses a data structure called the "segmented changelog" to represent its commit graph. See https://github.com/quark-zju/gitrevset/issues/1 and commit message at https://github.com/arxanas/git-branchless/commit/f6c540fea8392223d604c4994b081b603b3df850 for more details. It answers O(log(n)) merge-base queries with very good real performance (~1ms per query). That particular implementation is GPL-3-only, unfortunately.

**Reply by epage** [reply]
Ok, hadn't realized Eden was providing all of that functionality.

**Reply by epage** [reply]
https://github.com/gitext-rs/git-stack/pull/220 is merged

v0.8.2 will automatically split off these old branches before building the graph, speeding things up.

`git stack` is still about 6s in my use cases, so the gain were significant but not enough for frequent use.  I'm assuming we'd need on-disk caching for that.

**Reply by troglobit** [reply]
Amazing, will try it out tomorrow @ $DAYJOB \o/

**Reply by arxanas** [reply]
@epage for an on-disk merge-base cache, feel free to copy https://github.com/arxanas/git-branchless/blob/b48ed600777f73e410732e613b199a09e6dafaf2/src/core/mergebase.rs if it helps.

**Reply by epage** [reply]
I've created https://github.com/gitext-rs/git-stack/issues/222 to split out conversations on how to go about this

**Comment by arxanas** [comment]
@troglobit Can you try installing [git-branchless](https://github.com/arxanas/git-branchless) and running [`git move`](https://github.com/arxanas/git-branchless/wiki/Command:-git-move) on the same rebase? (Or install the latest version from source and try [`git sync`](https://github.com/arxanas/git-branchless/wiki/Command:-git-sync). If it completes substantially faster, then it's probably https://github.com/libgit2/libgit2/issues/6036.

EDIT: oops, I misunderstood which operation was slow, so I think the above doesn't apply.

@epage assuming that it's the above issue, I would love to get git-branchless's implementations of cherry-picking, etc. into [git2-ext](https://github.com/gitext-rs/git2-ext). That being said, the stated goals of the project are to 'Provide "good enough" implementations'. Should we expand the scope of the project to provide high-performance/production-ready implementations?

PS: the "about" text for the `git2-ext` repository on the Github page is broken, it just says "git Exte" and is cut off.

Reference on git-branchless's `cherry_pick_fast`: https://github.com/arxanas/git-branchless/blob/a4bc3c0c31b7be5cdd95ed8cfe4385e384482a99/src/git/repo.rs#L914-L924

**Reply by epage** [reply]
> @epage assuming that it's the above issue, I would love to get git-branchless's implementations of cherry-picking, etc. into [git2-ext](https://github.com/gitext-rs/git2-ext). That being said, the stated goals of the project are to 'Provide "good enough" implementations'. Should we expand the scope of the project to provide high-performance/production-ready implementations?

Yes, just haven't gotten to this yet.

The "good enough" just means we're not going to try to support every single use case possible,. that it serves as a common case and documentation for customization.  If you have better ideas on conveying that, I'm game.

> PS: the "about" text for the git2-ext repository on the Github page is broken, it just says "git Exte" and is cut off.

I've updated that but for some reason the `settings` app isn't working

Oh right, forgot to install it in the new org.

**Reply by epage** [reply]
Broke the git2-ext conversations out into https://github.com/gitext-rs/git2-ext/issues/7
